File: converter/source_reader.py
import sqlite3
from pathlib import Path
from typing import Optional


class DatabaseError(Exception):
    pass


class ReadOnlyDatabase:
    def __init__(self, db_path: str):
        self.db_path = Path(db_path)
        self._conn: Optional[sqlite3.Connection] = None

    def open(self):
        if not self.db_path.exists():
            raise DatabaseError(f"Database not found: {self.db_path}")
        uri = f"file:{self.db_path}?mode=ro"
        try:
            self._conn = sqlite3.connect(uri, uri=True)
        except sqlite3.Error as e:
            raise DatabaseError(f"Failed to open database: {e}")

    def close(self):
        if self._conn:
            self._conn.close()
            self._conn = None

    def validate(self):
        if self._conn is None:
            raise DatabaseError("Database not open")
        try:
            cur = self._conn.execute("PRAGMA schema_version")
            version = cur.fetchone()[0]
            if version < 1:
                raise DatabaseError(f"Invalid schema version: {version}")
        except sqlite3.Error as e:
            raise DatabaseError(f"Failed to validate database: {e}")

    def query(self, sql: str, params=None):
        if self._conn is None:
            raise DatabaseError("Database not open")
        try:
            if params:
                return self._conn.execute(sql, params).fetchall()
            return self._conn.execute(sql).fetchall()
        except sqlite3.Error as e:
            raise DatabaseError(f"Query failed: {e}")

    def __enter__(self):
        self.open()
        self.validate()
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.close()
        return False

File: converter/source_reader.js
import Database from 'better-sqlite3';
import { accessSync, constants } from 'node:fs';

export class DatabaseError extends Error {
  constructor(message) {
    super(message);
    this.name = 'DatabaseError';
  }
}

export class ReadOnlyDatabase {
  constructor(dbPath) {
    this.dbPath = dbPath;
    this._conn = null;
  }

  open() {
    try {
      accessSync(this.dbPath, constants.R_OK);
    } catch {
      throw new DatabaseError(`Database not found: ${this.dbPath}`);
    }
    try {
      this._conn = new Database(this.dbPath, { readonly: true, fileMustExist: true });
    } catch (e) {
      throw new DatabaseError(`Failed to open database: ${e.message}`);
    }
  }

  close() {
    if (this._conn) {
      this._conn.close();
      this._conn = null;
    }
  }

  validate() {
    if (!this._conn) {
      throw new DatabaseError('Database not open');
    }
    try {
      const row = this._conn.pragma('schema_version', { simple: true });
      if (row < 1) {
        throw new DatabaseError(`Invalid schema version: ${row}`);
      }
    } catch (e) {
      if (e instanceof DatabaseError) throw e;
      throw new DatabaseError(`Failed to validate database: ${e.message}`);
    }
  }

  query(sql, params) {
    if (!this._conn) {
      throw new DatabaseError('Database not open');
    }
    try {
      const stmt = this._conn.prepare(sql);
      return params ? stmt.all(...params) : stmt.all();
    } catch (e) {
      throw new DatabaseError(`Query failed: ${e.message}`);
    }
  }

  openSync() {
    this.open();
    this.validate();
    return this;
  }

  [Symbol.dispose]() {
    this.close();
  }
}
